src/db_reader.rs

```diff
@@ -0,0 +1,1099 @@
+use crate::bytes_range::BytesRange;
+use crate::config::{
+    CheckpointOptions, Clock, DbReaderOptions, ReadOptions, ScanOptions, SystemClock,
+};
+use crate::db_reader::ManifestPollerMsg::Shutdown;
+use crate::db_state::CoreDbState;
+use crate::db_stats::DbStats;
+use crate::error::SlateDBError;
+use crate::manifest::store::{ManifestStore, StoredManifest};
+use crate::manifest::Manifest;
+use crate::mem_table::{ImmutableMemtable, ImmutableWal, KVTable};
+use crate::reader::{ReadSnapshot, Reader};
+use crate::sst::SsTableFormat;
+use crate::sst_iter::SstIteratorOptions;
+use crate::stats::StatRegistry;
+use crate::tablestore::TableStore;
+use crate::utils::MonotonicClock;
+use crate::wal_replay::{WalReplayIterator, WalReplayOptions};
+use crate::{utils, Checkpoint, DbIterator};
+use bytes::Bytes;
+use log::{info, warn};
+use object_store::path::Path;
+use object_store::ObjectStore;
+use once_cell::sync::Lazy;
+use parking_lot::{Mutex, RwLock};
+use std::collections::VecDeque;
+use std::ops::{RangeBounds, Sub};
+use std::sync::Arc;
+use tokio::runtime::Handle;
+use tokio::select;
+use tokio::sync::mpsc::{UnboundedReceiver, UnboundedSender};
+use uuid::Uuid;
+
+/// Read-only interface for accessing a database from either
+/// the latest persistent state or from an arbitrary checkpoint.
+pub struct DbReader {
+    inner: Arc<DbReaderInner>,
+    manifest_poller: Option<ManifestPoller>,
+}
+
+struct DbReaderInner {
+    manifest_store: Arc<ManifestStore>,
+    table_store: Arc<TableStore>,
+    options: DbReaderOptions,
+    state: Arc<RwLock<Arc<CheckpointState>>>,
```

Comment:
this can just be `RwLock<Arc<CheckpointState>>` since `state` itself isn't shared (it's always accessed via `DbReaderInner`)